Add `Filterable` to the `IsEvent` constraints

i-am-tom opened the pull request:
As requested, the PR has been updated. I closed down the others, and I'll sort them out before reopening. Hopefully, this adds a `Filterable` instance to both `Event` and `Semantic`!

Commit:
Implement `Filterable`, add `mapWithAccum`+`when`

This commit adds two new functions to the library: `mapWithAccum`, a
function to help map over streams with respect to an accumulating value,
and `when`, which allows you to filter streams by other streams.

As well as this, the Filterable class has been added to replace a couple
of other utilities, such as `filter` and `mapMaybe`.

Commit:
Merge branch 'master' of github.com:paf31/purescript-behaviors

Merging in updates and altering PR to match.

paf31 commented on src/FRP/Event.purs:
```diff
@@ -42,6 +44,18 @@ foreign import never :: forall a. Event a
+
+  partition p xs = { yes: filter p xs, no: filter (not <<< p) xs }
+
+  partitionMap f xs = let ys = f <$> xs in
```
I think we can define this a little more succinctly in terms of `filterMap`.

i-am-tom replied:
Yeah, I think I was trying to be clever and save a call, but it actually just ends up being a mess. Will do.

paf31 commented on src/FRP/Event/Class.purs:
```diff
@@ -4,7 +4,7 @@ module FRP.Event.Class
   , count
   , mapAccum
-  , mapMaybe
+  , module Data.Filterable
```
Could you please move this to the bottom of the list?

i-am-tom replied:
Of course!

paf31 commented:
👍 Thanks, looks great!

Commit:
Making changes outlined in the code review

The re-export of `Filterable` has been moved to the bottom of the
export list for `FRP.Event.Class`, and `partitionMap` has been
defined slightly more neatly in terms of `filterMap`.
